src/partition/ReceivedPartition.cpp

```diff
@@ -469,144 +470,345 @@ void ReceivedPartition::createOwnerInformation()
-        PRECICE_DEBUG("Rank {} has tags {}", rank, slaveTags[rank]);
-        PRECICE_DEBUG("Rank {} has global IDs {}", rank, slaveGlobalIDs[rank]);
-        bool atInterface = false;
-        utils::MasterSlave::_communication->receive(atInterface, rank);
-        if (atInterface)
-          ranksAtInterface++;
+    // to store receive requests.
+    std::vector<com::PtrRequest> vertexNumberRequests;
+
+    // Define and initialize for load balancing
+    std::map<int, int> neighborRanksVertexCount;
+    for (auto &neighborRank : localConnectedBBMap) {
+      neighborRanksVertexCount.insert(std::make_pair(neighborRank.first, 0));
+    }
+
+    // Asynchronous recieve number of owned vertices from neighbor ranks
+    for (auto &neighborRank : localConnectedBBMap) {
+      auto request = utils::MasterSlave::_communication->aReceive(neighborRanksVertexCount.at(neighborRank.first), neighborRank.first);
+      vertexNumberRequests.push_back(request);
+    }
+
+    // Synchronous send number of owned vertices to neighbor ranks
+    for (auto &neighborRank : localConnectedBBMap) {
+      utils::MasterSlave::_communication->send(counter, neighborRank.first);
+    }
+
+    // wait until aReceive is complete.
+    for (auto &rqst : vertexNumberRequests) {
+      rqst->wait();
+    }
+
+    // #5: Exchange list of shared vertices with the neighbor
+
+    // to store send requests.
+    std::vector<com::PtrRequest> vertexListRequests;
+
+    for (auto &receivingRank : sharedVerticesSendMap) {
+      int  sendSize = receivingRank.second.size();
+      auto request  = utils::MasterSlave::_communication->aSend(sendSize, receivingRank.first);
+      vertexListRequests.push_back(request);
+      if (sendSize != 0) {
+        auto request = utils::MasterSlave::_communication->aSend(receivingRank.second, receivingRank.first);
+        vertexListRequests.push_back(request);
       }
     }
 
-    // Decide upon owners,
-    PRECICE_DEBUG("Decide owners, first round by rough load balancing");
-    PRECICE_ASSERT(ranksAtInterface != 0);
-    int localGuess = _mesh->getGlobalNumberOfVertices() / ranksAtInterface; // Guess for a decent load balancing
-    // First round: every slave gets localGuess vertices
-    for (int rank : utils::MasterSlave::allRanks()) {
-      int counter = 0;
-      for (size_t i = 0; i < slaveOwnerVecs[rank].size(); i++) {
-        // Vertex has no owner yet and rank could be owner
-        if (globalOwnerVec[slaveGlobalIDs[rank][i]] == 0 && slaveTags[rank][i] == 1) {
-          slaveOwnerVecs[rank][i]                 = 1; // Now rank is owner
-          globalOwnerVec[slaveGlobalIDs[rank][i]] = 1; // Vertex now has owner
-          counter++;
-          if (counter == localGuess)
-            break;
-        }
+    for (auto &neighborRank : sharedVerticesSendMap) {
+      int receiveSize = 0;
+      utils::MasterSlave::_communication->receive(receiveSize, neighborRank.first);
+      if (receiveSize != 0) {
```

Comment:
Isn't this already covered by the vector overload of receive?

Reply:
You mean the message size? I think not, we need to check to avoid zero-sized messages.